cabal-install/Distribution/Client/ProjectOrchestration.hs

```diff
@@ -243,14 +244,33 @@ runProjectBuildPhase verbosity ProjectBuildContext {..} =
 -- | Adjust an 'ElaboratedInstallPlan' by selecting just those parts of it
 -- required to build the given user targets.
 --
--- How to get the 'PackageTarget's from the 'UserBuildTarget' is customisable.
+-- How to get the 'PackageTarget's from the 'UserBuildTarget' is customisable,
+-- so that we can change the meaning of @pkgname@ to target a build or
+-- repl depending on which command is calling it.
 --
-selectTargets :: PackageTarget
+-- Conceptually, every target identifies one or more roots in the
+-- 'ElaboratedInstallPlan', which we then use to determine the closure
+-- of what packages need to be built, dropping everything from
+-- 'ElaboratedInstallPlan' that is unnecessary.
+--
+-- There are a few complications, however:
+--
+--      * Strangely (in ezyang's opinion) an ElaboratedConfiguredPackage
+--        doesn't actually dictate what should be built: we're asked
+--        to fill in 'pkgBuildTargets', 'pkgReplTarget' and
+--        'pkgBuildHaddock' to control what we actually do in the
+--        build phase.  This means we have to fiddle around with
+--        the ElaboratedConfiguredPackage roots to say what it will
+--        build.
```

Comment:
Well, naively, everything you could possibly want to build would be represented as a node in the graph; so selecting targets would involve ONLY picking the roots to keep, taking their transitive closure, and building this.

Obviously we can't actually do this in practice, because it isn't possible (or desirable) to invoke `Setup` scripts to build each file individually. So instead we end up with this weird compromise which is how it is done today. Even then, I'm not completely convinced by the current data structure: arguably `pkgBuildTargets` and friends should be stored out-of-band. But I didn't want to touch it. I can probably reword the comment to be clearer.